kore/src/Kore/Builtin/Set.hs

```diff
@@ -658,65 +782,24 @@ addToSetDisjoint = foldM addElementDisjoint
         then Nothing
         else return (Set.insert element set)
 
-{- | Transforms a normalized set into a @TermLike@ representation.
+{- | Computes the union of two sets if they are disjoint. Returns @Nothing@
+otherwise.
 -}
-normalizedToTerm
-    :: forall variable
-    .   ( Ord variable
-        , SortedVariable variable
-        , Unparse variable
-        )
-    => SmtMetadataTools Attribute.Symbol
-    -> Sort
-    -> NormalizedSet variable
-    -> TermLike variable
-normalizedToTerm
-    tools
-    sort1
-    NormalizedSet
-        { elementsWithVariables
-        , concreteElements
-        , sets
-        }
-  =
-    if Set.null concreteElements
-    then
-        case filteredSets of
-            [] -> case Set.toList elementsWithVariables of
-                [] -> asInternal tools sort1 concreteElements
-                (ewv : ewvs) -> addElements (element ewv) ewvs
-            (set:sets1) ->
-                let base = addSets set sets1
-                in addElements base (Set.toList elementsWithVariables)
-    else
-        let baseC = asInternal tools sort1 concreteElements
-            baseS = addSets baseC filteredSets
-        in addElements baseS (Set.toList elementsWithVariables)
+addToListDisjoint
+    :: (Ord a, Traversable t)
+    => Set a
+    -> [a]
+    -> t a
+    -> Maybe [a]
+addToListDisjoint set1 list1 list2 = do
+    (listResult, _) <- foldM addElementDisjoint (list1, set1) list2
+    return listResult
   where
-    addElements :: TermLike variable -> [TermLike variable] -> TermLike variable
-    addElements = List.foldr (\elem1 term -> concat' (element elem1) term)
-
-    addSets :: TermLike variable -> [TermLike variable] -> TermLike variable
-    addSets = List.foldr concat'
-
-    filteredSets :: [TermLike variable]
-    filteredSets = filter (not . isEmptySet) (Set.toList sets)
-
-    isEmptySet :: TermLike variable -> Bool
-    isEmptySet
-        (Builtin_ (Domain.BuiltinSet Domain.InternalSet { builtinSetChild }))
-      = Set.null builtinSetChild
-    isEmptySet (App_ symbol _)
-      | isSymbolUnit hookTools symbol = True
-      | otherwise = False
-    isEmptySet _ = False
-
-    elementSymbol = Builtin.lookupSymbolElement tools sort1
-    concatSymbol = Builtin.lookupSymbolConcat tools sort1
-    element elem' = mkApplySymbol elementSymbol [elem']
-    concat' set1 set2 = mkApplySymbol concatSymbol [set1, set2]
-
-    hookTools = Attribute.Symbol.hook <$> tools
+    addElementDisjoint :: Ord a => ([a], Set a) -> a -> Maybe ([a], Set a)
```

Comment:
I might be mistaking, but I think I saw this function in a different where clause. Would it be worth it to make it a helper function in the module that both functions could reuse?

Reply:
Fixed.